### PR TITLE
feat: Erc20 escrow payment processor (complete w/tests).

### PR DESCRIPTION
**Pull request by Dadogg80**
Feat: Payment-processing for erc20EscrowToPay smart-contract. 

-  IMPORTANT, please review [PR#624](https://github.com/RequestNetwork/requestNetwork/pull/624) first.

### CODE REVIEW

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+      // Payer initiate emergency claim.
+      await (await initiateEmergencyClaim(request, payee)).wait(1);
+      
+      // Payer reverts the emergency claim.
+      const tx = await revertEmergencyClaim(request, wallet);
+      const confirmedTx = await tx.wait(1);
+
+      // Checks the status and tx.hash.
+      expect(confirmedTx.status).toBe(1);
+      expect(tx.hash).not.toBeUndefined();
```
Comment (yomarion) [inline packages/payment-processor/test/payment/erc20-escrow-payment.test.ts:298]:
```suggestion
      expect(tx.hash).toBeDefined();
```
Easier to read :)

There are a few occurrences like this.

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+      request.requestId = 'aaff';
+
+      // Execute payEscrow.
+      await (await payEscrow(request, wallet, undefined, undefined)).wait(1);
+
+      // Payer executes a freeze of escrow funds.
+      await (await freezeRequest(request, wallet)).wait(1);
+      
+      // Expects the transaction to revert.
+      await expect(refundFrozenFunds(request, wallet)).rejects.toThrowError();
```
Comment (yomarion) [inline packages/payment-processor/test/payment/erc20-escrow-payment.test.ts:328]: > // Expects the transaction to revert.

Explain what you test instead, don't rephrase the code

```suggestion
      await expect(refundFrozenFunds(request, wallet)).rejects.toThrowError(
        'here the error you expect',
      );
```

That's to be sure we don't throw for buggy test mocks reasons, for instance.

**Review by yomarion — APPROVED** [review]
Well done ! Good job

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts (hunk 2 above)
Comment (Dadogg80) [reply packages/payment-processor/test/payment/erc20-escrow-payment.test.ts:328]: Thank you, I'll fix that.

File: packages/payment-processor/src/payment/erc20-escrow-payment.ts [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    data: encodedTx,
+    to: contractAddress,
+    value: 0,
+    ...overrides,
+  });
+  return tx;
+}
+
+/**
+ * Processes a transaction to refundFrozenFunds().
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/payment-processor/src/payment/erc20-escrow-payment.ts:213]: I agree, I added it more for testing purposes. 
I'll remove it.

[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/payment-processor/src/payment/erc20-escrow-payment.ts:213]: I removed the `requestMapping` from here and `erc20-escrow-payment.test.ts`.

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts (hunk 2 above)
Comment (Dadogg80) [reply packages/payment-processor/test/payment/erc20-escrow-payment.test.ts:328]:
```suggestion
      // Payer tries to withdraw frozen funds before unlock date.
      await expect(refundFrozenFunds(request, wallet)).rejects.toThrowError('Not Yet!',);
```

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts (hunk 1 above)
Comment (Dadogg80) [reply packages/payment-processor/test/payment/erc20-escrow-payment.test.ts]: Updated all occurrences. :-)

File: packages/payment-processor/test/payment/erc20-escrow-payment.test.ts [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        `0xa58ad6bc00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000008${values.paymentReference}000000000000000000000000000000000000000000000000`,
+      );
+    });
+  });
+
+  describe('Test function calls:', () => {
+    beforeEach(async () => {
+      await approveErc20ForEscrow(validRequest, erc20ContractAddress, wallet);
+    });
+    it('Should pay the amount and fee from payers account', async () => {
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/payment-processor/test/payment/erc20-escrow-payment.test.ts:198]: Can we talk about this?

[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/payment-processor/test/payment/erc20-escrow-payment.test.ts]: Done! Resolving conversation

File: packages/payment-processor/src/payment/erc20-escrow-payment.ts [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
+ */
+export function encodePayEscrow(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: providers.Web3Provider | Signer = getProvider(),
+  amount?: BigNumberish,
+  feeAmountOverride?: BigNumberish,
+): string {
+  validateRequest(request, PaymentTypes.PAYMENT_NETWORK_ID.ERC20_FEE_PROXY_CONTRACT);
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/erc20-escrow-payment.ts:222]: I'm not yet super familiar with our Escrow solution... isn't the escrow a new Payment Network?

File: packages/payment-processor/src/index.ts [hunk 6]
```diff
@@ -12,6 +12,7 @@ export * from './payment/swap-any-to-erc20';
 export * from './payment/swap-erc20';
 export * from './payment/swap-erc20-fee-proxy';
 export * from './payment/conversion-erc20';
+export * from './payment/erc20-escrow-payment';
```
Comment (benjlevesque) [inline packages/payment-processor/src/index.ts:15]: considering the number of functions and the specificities, I would maybe group these exports like this: 
```suggestion
export * as Escrow from './payment/erc20-escrow-payment';
```

so the usage would be 

```ts 
import { Escrow } from "@requestnetwork/payment-processor";

Escrow.approveErc20ForEscrow(...)

```

**Review by benjlevesque — APPROVED** [review]

File: packages/payment-processor/src/payment/erc20-escrow-payment.ts (hunk 5 above)
Comment (Dadogg80) [reply packages/payment-processor/src/payment/erc20-escrow-payment.ts:222]: No, it's using the erc20_fee_proxy.

File: packages/payment-processor/src/index.ts (hunk 6 above)
Comment (Dadogg80) [reply packages/payment-processor/src/index.ts:15]: Thanx, I fix.

**Comment by Dadogg80** [comment]
This PR #675 has been squashed and merged to branch [escrow-payment-processor-v1](https://github.com/RequestNetwork/requestNetwork/pull/693), new PR #693 is created. 

Closing PR #675.

**Pull request closed without merging**